### PR TITLE
Better errors when rustc cannot derive lifetimes

### LINKED ISSUE #30086: "0 elided lifetimes" is confusing
```rust
fn a() -> &i32 {
    panic!();
}
fn b(x: i32) -> &i32 {
    panic!();
}
```

```
test.rs:1:11: 1:15 help: this function's return type contains a borrowed value,
but there is no value for it to be borrowed from
test.rs:4:17: 4:21 help: this function's return type contains a borrowed value,
but the signature does not say which one of `x`'s 0 elided lifetimes it is borrowed from
```

Help for `a` is reasonable, but help for `b` is not. See also #22760.

### PR DESCRIPTION
**Pull request by jFransham**
Fixes #30086

### CODE REVIEW

**Comment by Manishearth** [comment]
@bors r+ force

Thanks!

**Comment by sanxiyn** [comment]
This needs a test.

**Comment by Manishearth** [comment]
Wait, sorry, we should add a test for this. Could you add one to `tests/compile-fail` which verifies that the error messages are shown correctly? Check out the other tests in that folder (or ask me on IRC) for pointers on how to write one.

File: src/librustc_typeck/astconv.rs [hunk 1]
```diff
@@ -229,17 +233,26 @@ fn report_elision_failure(
...
-    } else if len == 0 {
+
+    if len == 0 {
         fileline_help!(tcx.sess, default_span,
                        "this function's return type contains a borrowed value, but \
                         there is no value for it to be borrowed from");
         fileline_help!(tcx.sess, default_span,
                        "consider giving it a 'static lifetime");
+    } else if !any_lifetimes {
+        fileline_help!(tcx.sess, default_span,
```
Comment (arielb1) [inline src/librustc_typeck/astconv.rs]: Why is the error message different between this case and the previous one?

Comment (jFransham) [inline src/librustc_typeck/astconv.rs]: I was thinking the previous message implied that there were no arguments to draw from, and I was hoping to make the message as explicit as possible; I think if they were made the same they would be slightly more confusing. I don't feel strongly on this or anything, I'd be happy to change it so they have the same message if that's the consensus.

Also yeah, I'll do the tests.

**Comment by Manishearth** [comment]
@bors r+

Sorry, didn't notice the new commit.

Thanks!

**Pull request merged**